src/v/security/role_store.h

```diff
@@ -0,0 +1,159 @@
+#include "absl/container/node_hash_map.h"
+#include "absl/container/node_hash_set.h"
+#include "security/role.h"
+#include "security/types.h"
+
+#include <absl/algorithm/container.h>
+#include <boost/range/iterator_range.hpp>
+
+#include <ranges>
+#include <string_view>
+#include <type_traits>
+
+namespace security {
+
+/*
+ * Store for roles and role members.
+ *
+ * This store is subject to two distinct usage cases:
+ *   1. Retrieve the list of roles assigned to some user
+ *      - Used inline as part of the request authorization flow
+ *   2. Read and write from  the set of all roles
+ *      - Used for servicing role management requests in the Admin API
+ *
+ * The internal structure of role_store is geared toward servicing
+ * (1) as efficiently as possible with the primary goal of avoiding
+ * performance regression in the authorizer.
+ *
+ */
+class role_store {
+    struct role_name_view
+      : public named_type<std::string_view, struct role_name_view_tag> {
+        explicit role_name_view(const role_name& r)
+          : named_type<std::string_view, struct role_name_view_tag>(r()) {}
+    };
+    using role_set_type = absl::node_hash_set<role_name>;
+    using name_view_set_type = absl::flat_hash_set<role_name_view>;
+    using members_store_type
+      = absl::node_hash_map<role_member, name_view_set_type>;
+
+public:
+    role_store() noexcept = default;
+    role_store(const role_store&) = delete;
+    role_store& operator=(const role_store&) = delete;
+    role_store(role_store&&) noexcept = default;
+    role_store& operator=(role_store&&) noexcept = default;
+    ~role_store() noexcept = default;
+
+    // Add the given role and members range iff it is not already present.
+    template<typename T>
+    requires std::ranges::range<T>
+             && std::convertible_to<std::ranges::range_value_t<T>, role_member>
+    bool put(role_name name, T&& role) {
+        auto [it, inserted] = _roles.insert(std::move(name));
+        if (inserted) {
+            for (const auto& m : role) {
+                _members_store[m].insert(role_name_view{*it});
+            }
+        }
+        return inserted;
+    }
+
+    std::optional<role> get(const role_name& name) const {
+        if (!_roles.contains(name)) {
+            return std::nullopt;
+        }
+        auto member_rng = _members_store
+                          | std::views::filter(
+                            [&name](const members_store_type::value_type& e) {
+                                return e.second.contains(role_name_view{name});
+                            })
+                          | std::views::keys;
+        return role{{member_rng.begin(), member_rng.end()}};
+    }
+
+    boost::iterator_range<name_view_set_type::const_iterator>
+    roles_for_member(const role_member& user) const {
+        if (auto it = _members_store.find(user); it != _members_store.end()) {
+            return it->second;
+        }
+        return {};
+    }
```

Comment:
question: Should there be any assert checks to ensure that the roles that the member belongs to are also in `_roles`?  I don't see this situation happening as the code looks solid, but just in case a role is removed from `_roles` but not from the `_members_store`.